bottlecap/src/lifecycle/invocation/triggers/kinesis_event.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+impl Trigger for KinesisEventRecord {
+    fn new(payload: Value) -> Option<Self> {
+        let records = payload.get("Records").and_then(Value::as_array);
+        match records {
+            Some(records) => match serde_json::from_value::<KinesisEventRecord>(records[0].clone())
+            {
+                Ok(event) => Some(event),
+                Err(e) => {
+                    debug!("Failed to deserialize Kinesis Record: {e}");
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+
+    fn is_match(payload: &Value) -> bool {
+        if let Some(first_record) = payload
+            .get("Records")
+            .and_then(Value::as_array)
+            .and_then(|r| r.first())
+            .take()
+        {
+            first_record.get("kinesis").is_some()
+        } else {
+            false
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    fn enrich_span(&self, span: &mut Span) {
+        let event_source_arn = &self.event_source_arn;
+        let split_source_arn: Vec<&str> = event_source_arn.split('/').collect();
+        let parsed_stream_name = split_source_arn.last().unwrap_or(&"");
+        let parsed_shard_id = self.event_id.split(':').next().unwrap_or_default();
+        let service_name = "kinesis";
+        span.name = "aws.kinesis".to_string();
+        span.service = service_name.to_string();
+        span.start = (self.kinesis.approximate_arrival_timestamp * MS_TO_NS) as i64;
+        span.resource = (*parsed_stream_name).to_string();
+        span.r#type = "web".to_string();
+        span.meta = HashMap::from([
+            ("operation_name".to_string(), "aws.kinesis".to_string()),
+            (
+                "resource_names".to_string(),
+                (*parsed_stream_name).to_string(),
+            ),
+            ("stream_name".to_string(), (*parsed_stream_name).to_string()),
+            ("shard_id".to_string(), parsed_shard_id.to_string()),
+            ("event_source_arn".to_string(), event_source_arn.to_string()),
+            ("event_id".to_string(), self.event_id.to_string()),
+            ("event_name".to_string(), self.event_name.to_string()),
+            ("event_version".to_string(), self.event_version.to_string()),
+            (
+                "partition_key".to_string(),
+                self.kinesis.partition_key.to_string(),
+            ),
+        ]);
+    }
+
+    fn get_tags(&self) -> HashMap<String, String> {
+        HashMap::from([
+            (
+                FUNCTION_TRIGGER_EVENT_SOURCE_TAG.to_string(),
+                "kinesis".to_string(),
+            ),
+            (
+                FUNCTION_TRIGGER_EVENT_SOURCE_ARN_TAG.to_string(),
+                self.event_source_arn.clone(),
+            ),
+        ])
+    }
+
+    fn get_arn(&self, _region: &str) -> String {
+        self.event_source_arn.clone()
+    }
+
+    fn get_carrier(&self) -> HashMap<String, String> {
+        HashMap::new()
```

Comment:
There is an actual carrier for Kinesis tho?

Reply:
there is nothing in the json example, I will try to find it in the test